jupyter_server/services/kernels/kernelmanager.py

```diff
@@ -711,4 +715,103 @@ class ServerKernelManager(AsyncIOLoopKernelManager):
+    # custom event schemas.
+    extra_event_schema_paths = List(
+        default_value=[],
+        help="""
+        A list of pathlib.Path objects pointing at to register with
+        the kernel manager's eventlogger.
+        """,
+    ).tag(config=True)
+
+    event_logger = Instance(EventLogger)
+
+    @default("event_logger")
+    def _default_event_logger(self):
+        """Initialize the logger and ensure all required events are present."""
+        if self.parent is not None and hasattr(self.parent, "event_logger"):
+            logger = self.parent.event_logger
+        else:
+            # If parent does not have an event logger, create one.
+            logger = EventLogger()
+        # Ensure that all the expected schemas are registered. If not, register them.
+        schemas = self.core_event_schema_paths + self.extra_event_schema_paths
+        for schema_path in schemas:
+            # Try registering the event.
+            try:
+                logger.register_event_schema(schema_path)
+            # Pass if it already exists.
+            except SchemaRegistryException:
+                pass
+        return logger
+
+    def emit(self, schema_id, data):
+        """Emit an event from the kernel manager."""
+        self.event_logger.emit(schema_id=schema_id, data=data)
+
+    @wraps(AsyncIOLoopKernelManager.start_kernel)
+    async def start_kernel(self, *args, **kwargs):
+        result = await super().start_kernel(*args, **kwargs)
+        self.emit(
+            schema_id="https://events.jupyter.org/jupyter_server/kernel_actions/v1",
+            data={
+                "kernel_id": self.kernel_id,
+                "kernel_name": self.kernel_name,
+                "action": "start",
+                "msg": f"Kernel {self.kernel_id} was started.",
+            },
+        )
+        return result
+
+    @wraps(AsyncIOLoopKernelManager.interrupt_kernel)
+    async def interrupt_kernel(self, *args, **kwargs):
+        result = await super().interrupt_kernel(*args, **kwargs)
+        self.emit(
+            schema_id="https://events.jupyter.org/jupyter_server/kernel_actions/v1",
+            data={
+                "kernel_id": self.kernel_id,
+                "kernel_name": self.kernel_name,
+                "action": "interrupt",
+                "msg": f"Kernel {self.kernel_id} was interrupted.",
+            },
+        )
+        return result
+
+    @wraps(AsyncIOLoopKernelManager.shutdown_kernel)
+    async def shutdown_kernel(self, *args, **kwargs):
+        result = await super().shutdown_kernel(*args, **kwargs)
+        self.emit(
+            schema_id="https://events.jupyter.org/jupyter_server/kernel_actions/v1",
+            data={
+                "kernel_id": self.kernel_id,
+                "kernel_name": self.kernel_name,
+                "action": "shutdown",
+                "msg": f"Kernel {self.kernel_id} was shutdown.",
+            },
+        )
```

Comment:
This is good, however, it doesn't emit exceptions from super. You may need to catch exception, emit and then raise.

Reply:
Since we have to duplicate the emit calls here and in Gateway's Kernel Manager, I think its important that the content for the events be identical - same data and message formats - otherwise event-processing applications will have a frustrating time having to deal with two different "fingerprints".

Reply:
The new decorator I created, `emit_kernel_action_event`, eliminates the need to maintain duplicates of these events.